### ISSUE TITLE
MIR borrowck: errors unreported in unreachable arms of match expressions

### ISSUE THREAD

**Issue by zilbuz** [rust-lang/rust#45043]
*Note from @pnkfelix for people looking at this bug: See important soundness issue in [ariel's issue](https://github.com/rust-lang/rust/issues/45045).*

In the following code the borrow error in the `z` arm isn't reported with MIR borrowck, but it is reported with AST borrowck:

```rust
fn main() {
    let mut x = 1;
    let _x = &mut x;
    match x {
        y => println!("y: {}", y), // Error in AST and MIR
        z => println!("z: {}", z), // Error only in AST
    }
}
```
(compile with `-Z emit-end-regions -Z borrowck-mir` to use the MIR borrowck)

**Comment by arielb1** [comment]
This is actually an unsoundness in AST borrowck: AST borrowck doesn't check things that are matched on for conflicting borrows unless there are actually pattern bindings, e.g. this compiles:

```Rust
enum Xyz {
    A,
    B,
}

fn main() {
    let mut e = Xyz::A;
    let f = &mut e;
    match e {
        Xyz::A => println!("a"),
        Xyz::B => println!("b"),
    };
    *f = Xyz::B;
}
```

That is unsound because of e.g. data races. For example, this code compiles and runs, and semi-reliably segfaults:
```Rust
#![feature(test)]
use std::{thread, time};

extern crate crossbeam;
extern crate test;

enum Xyz<'a> {
    A(&'a usize),
    B(usize),
}

fn main() {
    let mut e = Xyz::A(&0);
    crossbeam::scope(|scope| {
        scope.spawn(|| {
            let now = time::Instant::now();
            let ten_millis = time::Duration::from_millis(50);
            while now.elapsed() < ten_millis {
                for _ in 0..1000000 {
                    e = Xyz::A(&0);
                    test::black_box(()); // compiler barrier
                    e = Xyz::B(0xaaaaaaaa);
                }
            }
        });
        let ten_millis = time::Duration::from_millis(10);
        thread::sleep(ten_millis);
        match e {
            Xyz::A(&0) => println!("a"),
            _ => println!("b"),
        };
    });
}
```

**Comment by mikhail-m1** [comment]
my idea of fix https://github.com/mikhail-m1/rust/commit/ff15f84c0259da3dc90266a43b9db91ccc8d537a, it's for but not finished
